### DISCUSSION TITLE
Could caffeine support time-based expiration specific for each entry

### DISCUSSION THREAD

**Discussion by waterlx** [ben-manes/caffeine#1741] [Q&A] [answered]
I read the wiki to decribe [time-based expiration](https://github.com/ben-manes/caffeine/wiki/Eviction#time-based).
If I get that correctly, the expiration time applies to all entries in the cache.
I am just trying to check if different expiration time could be set to each entry. Is "Evict based on a varying expiration policy" designed for that?

I am new to this. Thanks in advance!

**Comment by ben-manes** [comment, marked as answer]
The per-entry policy is defined using the Expiry interface. This lets you determine the entry’s duration based on its key, value, and operation. 

For example if the http response headers provided a TTL when loading a value into the cache, the policy could honor it locally to match this externally defined duration.

**Reply by ben-manes** [reply]
I suppose you might have been looking for a [put(key, value, duration)](https://www.javadoc.io/doc/com.github.ben-manes.caffeine/caffeine/latest/com.github.benmanes.caffeine/com/github/benmanes/caffeine/cache/Policy.VarExpiration.html) method. While that is available through cache.policy(), you still need an Expiry for anything loaded through the cache (which is highly preferable to avoid “cache stampedes”). Those apis are available when helpful, we just don’t emphasize them as often unnecessary or encourage bad practices.
